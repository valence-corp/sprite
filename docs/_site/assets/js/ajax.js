async function getDocument(url) {
  try {
    const response = await fetch(url);
    const html = await response.text();
    const parser = new DOMParser();
    const htmlDoc = parser.parseFromString(html, "text/html");
    const contentDiv = htmlDoc.getElementById("content");
    const contentHtml = contentDiv.outerHTML;
    document.getElementById("content").innerHTML = contentDiv.innerHTML;
  } catch (error) {
    throw new Error("Could not fetch page.", { cause: error });
  }
}

function createLinkClickEvents(root) {
  const elements = root.querySelectorAll("a");
  console.log(elements);
  if (elements.length) {
    elements.forEach((link) => {
      link.addEventListener("click", (e) => {
        e.preventDefault();
        getDocument(e.currentTarget.href);
      });
    });
  }
}

function ajax() {
  createLinkClickEvents(document);
}

document.addEventListener("DOMContentLoaded", ajax);
