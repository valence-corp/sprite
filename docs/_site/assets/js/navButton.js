document.addEventListener("DOMContentLoaded", function () {
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const mainNav = document.getElementById("mainNavigation");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let previousTouchPosition = null;
  let currentTranslation = 0;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation > -80) {
      currentTranslation = 0;
      mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;
    } else {
      mainNav.style.transform = "";
      currentTranslation = 0;
      closeMenu();
    }
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
