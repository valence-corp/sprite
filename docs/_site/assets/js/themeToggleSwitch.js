// function setTheme(theme) {
//   localStorage.setItem("theme", theme);
//   document.body.className = theme;
// }

// function setSwitchPosition(theme) {
//   switch (theme) {
//     case "dark":
//       themeToggleSwitch.checked = true;
//       break;
//     case "light":
//       themeToggleSwitch.checked = false;
//       break;
//     default:
//       throw new Error("unknow theme");
//   }
// }

// function setThemeAndSwitch(theme) {
//   setTheme(theme);
//   setSwitchPosition(theme);
// }

// function prefersTheme(theme) {
//   return window.matchMedia(`(prefers-color-scheme: ${theme})`).matches();
// }

// function initializeThemeToggle() {
//   const theme = localStorage.getItem("theme");
//   const toggleSwitch = document.getElementById("themeToggle");

//   toggleSwitch.addEventListener("click", () => {
//     if (themeToggleSwitch.checked) {
//       setTheme("dark");
//     } else {
//       setTheme("light");
//     }
//   });

//   if (theme) {
//     setThemeAndSwitch(theme);
//   } else {
//     if (prefersTheme("dark")) {
//       setThemeAndSwitch("dark");
//       return;
//     } else {
//       setThemeAndSwitch("light");
//     }
//   }
// }

// document.addEventListener("DOMContentLoaded", initializeThemeToggle);

class ThemeController {
  constructor(target) {
    this.element = target;
    this.theme = localStorage.getItem("theme");
    this.initialize();
  }
  initialize = () => {
    this.element.addEventListener("click", () => {
      if (this.element.checked) {
        this.changeTheme("dark");
      } else {
        this.changeTheme("light");
      }
    });
    this.determineInitialPosition();
  };
  prefersTheme = (theme) => {
    return window.matchMedia(`(prefers-color-scheme: ${theme})`).matches();
  };
  determineInitialPosition = () => {
    if (this.theme) {
      this.setTheme(this.theme);
    } else {
      if (prefersTheme("dark")) {
        this.setTheme("dark");
        return;
      } else {
        this.setTheme("light");
      }
    }
  };
  setPosition = (theme) => {
    switch (theme) {
      case "dark":
        this.element.checked = true;
        break;
      case "light":
        this.element.checked = false;
        break;
      default:
        throw new Error("unknow theme");
    }
  };
  changeTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
  };
  setTheme = (theme) => {
    this.changeTheme(theme);
    this.setPosition(theme);
  };
}

function initializeThemeToggle() {
  const element = document.getElementById("themeToggle");
  return new ThemeController(element);
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
