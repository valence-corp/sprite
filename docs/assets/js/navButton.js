document.addEventListener("DOMContentLoaded", function () {
  const button = document.getElementById("navButton");
  const desktopButton = document.getElementById("navToggleDesktop");
  const indicator = button.querySelector("svg");
  const nav = document.getElementById("mainNavigation");
  let open = false;

  desktopButton.addEventListener("click", function () {
    nav.classList.toggle("open");
  });

  button.addEventListener("click", function () {
    nav.classList.toggle("open");
    open = !open;
    if (open) {
      indicator.innerHTML = `
        <line x1="34" y1="18" x2="18" y2="34"></line>
        <line x1="18" y1="18" x2="34" y2="34"></line>
      `;
    } else {
      indicator.innerHTML = `
        <line x1="34" y1="30" x2="18" y2="30"></line>
        <line x1="18" y1="22" x2="34" y2="22"></line>
      `;
    }
  });
});
