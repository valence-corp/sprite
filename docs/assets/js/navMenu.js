document.addEventListener("DOMContentLoaded", function () {
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = "";
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
