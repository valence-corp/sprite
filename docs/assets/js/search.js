"use strict";

import data from "/sprite/assets/js/searchData.js";
import serializedIndex from "./serialized.js";

async function initializeSearch() {
  const config = {
    minSearchLength: 3,
    indexedFields: ["name", "desc", "type", "parent"],
  };

  // Build the index using Lunr.js
  const index = lunr.Index.load(serializedIndex);

  // Define the root element
  const root = document.getElementById("searchBox");

  // State definition
  const state = {
    filters: {
      class: true,
      method: true,
      type: true,
    },
    results: {
      filtered: [],
      full: [],
    },
    visible: false,
  };

  // Elements for reference
  const el = {
    input: root.querySelector("#searchInput"),
    results: root.querySelector("#searchResultsBox"),
    list: root.querySelector("#resultList"),
    numberOfResults: root.querySelector("#numberOfResults"),
    numberResultsShowing: root.querySelector("#numberResultsShowing"),
    backdrop: undefined,
    filters: root.querySelector("#searchFilters").getElementsByTagName("input"),
    searchLabel: root.querySelector("#searchLabel"),
    close: root.querySelector("#searchClose"),
  };

  // Initialize search filter toggles
  for (let i = 0; i < el.filters.length; ++i) {
    el.filters[i].addEventListener("change", () => {
      state.filters[el.filters[i].value] = el.filters[i].checked;
      updateResultList();
    });
  }

  function appendBackdrop() {
    el.backdrop = document.createElement("div");
    el.backdrop.id = "backdrop";
    el.backdrop.addEventListener("click", hideSearchBox);
    root.appendChild(el.backdrop);
  }

  function removeBackdrop() {
    el.backdrop.removeEventListener("click", appendBackdrop);
    el.backdrop.remove();
    el.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = el.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === el.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              el.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection();
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    el.results.classList.remove("hidden");
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    el.results.classList.add("hidden");
    document.removeEventListener("keydown", showResultsBox);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    el.list.innerHTML = "";
    el.numberOfResults.innerHTML = state.results.full.length;
    el.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        // TODO: should look into this, is it necessary to find
        // the original document?
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");

          li.className = doc.type;

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          el.list.appendChild(li);
        }
      });

      el.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  // search the index
  function search() {
    // Hide the "Search" Label
    // (which functions as a placeholder)
    if (el.input.value.length > 0) {
      el.searchLabel.classList.add("hidden");
    } else {
      el.searchLabel.classList.remove("hidden");
    }

    // run a search if the search term is longer than
    // the minimum set in the config
    if (el.input.value.length >= config.minSearchLength) {
      const initial = index.search(searchInput.value);

      // if there are no results, try again with a wildcard,
      // and then yet again with two wildcards
      if (initial.length === 0) {
        const wildcardResults = index.search(`${searchInput.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${searchInput.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }

  el.input.addEventListener("input", search);
  el.close.addEventListener("click", hideSearchBox);
}

document.addEventListener("DOMContentLoaded", initializeSearch);
