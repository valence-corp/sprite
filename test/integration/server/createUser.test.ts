import { variables } from '../variables.js';
import { client as testClient } from './testClient.js';

const config = {
  username: variables.newUsername,
  password: variables.newPassword,
  databases: {
    [variables.databaseName]: '*'
  }
};

describe('SpriteServer.createUser', () => {
  it('should create a user', async () => {
    const created = await testClient.createUser(config);
    expect(created).toBe(true);
  });

  it('should propagate errors from the database', async () => {
    await expect(testClient.createUser(config)).rejects.toMatchSnapshot();
  });
});
