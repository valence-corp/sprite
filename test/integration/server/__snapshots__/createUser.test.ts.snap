// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteServer.createUser should propagate errors from the database 1`] = `[Error: Could not create user newUser. In database(s): SpriteIntegrationTesting]`;
